libraries/libfc/include/fc/crypto/common.hpp

```diff
@@ -178,7 +178,7 @@ namespace fc { namespace crypto {
       {}
 
       shim(data_type&& data)
-      :_data(forward<data_type>(data))
+         :_data(std::move(data))
```

Comment:
This one should be `std::forward<Data>` since it based on the class template type `Data`.

Reply:
Hum, no, forward just works for function templates, not for regular member functions of a templated class.

Reply:
the `fact_t_forward<int&> ff2(x);` call is calling the default copy constructor. Add `fact_t_forward(val_t &) = delete;`  to see.

Reply:
https://rextester.com/FUUAH38143

Reply:
https://rextester.com/YBWP43749

Reply:
Ah right, thanks. `std::forward` is for universal references and a class template type is set at time of object creation making it not universal. The member of the templated class is just a normal member at the point of use.

Reply:
Yes, exactly. I created a better example demonstrating this https://rextester.com/LBXE59514